File: src/com/easy/AddStrings_415.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class AddStrings_415 {
    // 2019年7月18日
    // Use sequential parsing
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 4 ms, faster than 38.26%
    // Memory Usage: 36.3 MB, less than 99.80%
    public String addStrings(String num1, String num2) {
        // same concept as LeetCode 67. Add Binary
        StringBuilder mBuilder = new StringBuilder();
        int mLength1 = num1.length();
        int mLength2 = num2.length();
        int mLength = Math.max(mLength1, mLength2);
        int mCarry = 0;
        for (int i = 0; i < mLength; i++) {
            int mTmp1 = (mLength1 > i) ? (num1.charAt(mLength1 - i - 1) - '0') : 0;
            int mTmp2 = (mLength2 > i) ? (num2.charAt(mLength2 - i - 1) - '0') : 0;
            mBuilder.insert(0, (mTmp1 + mTmp2 + mCarry) % 10);
            mCarry = (mTmp1 + mTmp2 + mCarry) / 10;
        }
        if (mCarry != 0) {
            mBuilder.insert(0, mCarry);
        }

        return mBuilder.toString();
    }
}

File: Golang_src/easy/addStrings_415.go
package easy

import "strconv"

// Use iterative:
// Time Complexity: O(max(m,n))
// Space Complexity:O(max(m,n))
// m : length of num1
// n : length of num2
// Runtime: 8 ms, faster than 44.51%
// Memory Usage: 7.1 MB, less than 7.14%
func addStrings(num1 string, num2 string) string {

	result := ""

	str1 := []rune(num1)
	str2 := []rune(num2)

	index1 := len(str1) - 1
	index2 := len(str2) - 1

	carry := 0

	for index1 >= 0 || index2 >= 0 || carry != 0 {
		if index1 >= 0 {
			carry += int(str1[index1] - '0')
			index1--
		}
		if index2 >= 0 {
			carry += int(str2[index2] - '0')
			index2--
		}
		result = strconv.Itoa(carry%10) + result
		carry /= 10

	}

	return result
}
